subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/DependencyInsightReportTask.java

```diff
@@ -367,18 +416,50 @@ private void printSection(StyledTextOutput out, Section extraDetail, int depth)
         }
 
         private void printVariantDetails(StyledTextOutput out, RenderableDependency dependency) {
-            List<ResolvedVariantResult> resolvedVariants = dependency.getResolvedVariants();
-            for (ResolvedVariantResult resolvedVariant : resolvedVariants) {
+            if (dependency.getResolvedVariants().isEmpty() && dependency.getAllVariants().isEmpty()) {
+                return;
+            }
+            Set<String> selectedVariantNames = dependency.getResolvedVariants()
+                .stream()
+                .map(ResolvedVariantResult::getDisplayName)
+                .collect(Collectors.toSet());
+            if (showAllVariants) {
+                out.println();
+                out.withStyle(Header).text("-------------------");
```

Comment:
`withStyle` creates a whole new `StyledTextOutput` each time is is called.  I was using it for changing style mid line and then auto-reverting to the original style.  You should probably just use `out.style(Style)` once if the whole header block is in header style, then set it back to `Style.Normal` afterwards.